File: ios/Runner/NodeRuntime/SimiChatNodeRuntime.swift
import Foundation

/// Owns the NodeMobile.framework instance shipped inside the iOS app.
/// Node runs on a private queue for the lifetime of the app process; Flutter
/// only observes its local HTTP health/SSE endpoint.
final class SimiChatNodeRuntime {
  static let channel = "top.simitalk.aichat/node_runtime"
  private let bundle: Bundle
  private let queue = DispatchQueue(label: "top.simitalk.aichat.node-runtime", qos: .utility)
  private let lock = NSLock()
  private var started = false
  private var running = false
  private var state = 0 // stopped, starting, running, crashed
  private var exitCode = 0
  private var restartCount = 0
  private var lastError = ""
  private var scriptPath = ""
  private let port = 37651

  init(bundle: Bundle = .main) {
    self.bundle = bundle
  }

  func start() -> [String: Any] {
    do {
      try prepareRuntimeFiles()
    } catch {
      lock.lock()
      state = 3
      lastError = error.localizedDescription
      lock.unlock()
      return info(running: false)
    }

    lock.lock()
    if started {
      let result = infoLocked(running: running || state == 1)
      lock.unlock()
      return result
    }
    started = true
    state = 1
    exitCode = 0
    lastError = ""
    let script = scriptPath
    lock.unlock()

    queue.async { [weak self] in
      guard let self else { return }
      let support = self.applicationSupportDirectory()
      let extensionRoot = support.deletingLastPathComponent().path
      setenv("HOME", support.path, 1)
      setenv("TMPDIR", FileManager.default.temporaryDirectory.path, 1)
      setenv("NODE_PATH", support.path, 1)
      setenv("MCP_RUNTIME_HOST", "127.0.0.1", 1)
      setenv("MCP_RUNTIME_PORT", String(self.port), 1)
      setenv("MCP_RUNTIME_WORKSPACE_ROOT", support.path, 1)
      setenv("MCP_RUNTIME_EXTENSION_ROOT", extensionRoot, 1)
      setenv("SIMICHAT_NODE_RUNTIME_KIND", "ios-embedded", 1)
      setenv("SIMICHAT_NODE_APP_MANAGED", "true", 1)

      let arguments = ["node", script]
      var cArguments = arguments.map { strdup($0) }
      self.lock.lock()
      self.running = true
      self.state = 2
      self.lock.unlock()
      let code = cArguments.withUnsafeMutableBufferPointer { buffer in
        simichat_node_start(Int32(arguments.count), buffer.baseAddress)
      }
      cArguments.forEach { free($0) }
      self.lock.lock()
      self.running = false
      self.started = false
      self.exitCode = Int(code)
      self.state = code == 0 ? 0 : 3
      if code != 0 { self.lastError = "embedded Node exited with code \(code)" }
      self.lock.unlock()
    }
    return info(running: true)
  }

  func status() -> [String: Any] {
    lock.lock()
    let result = infoLocked(running: running || state == 1)
    lock.unlock()
    return result
  }

  func stop() -> [String: Any] {
    var result = status()
    result["stopSupported"] = false
    return result
  }

  private func info(running: Bool) -> [String: Any] {
    lock.lock()
    let result = infoLocked(running: running)
    lock.unlock()
    return result
  }

  private func infoLocked(running: Bool) -> [String: Any] {
    [
      "running": running,
      "state": stateName(state),
      "nativeState": state,
      "nativeExitCode": exitCode,
      "restartCount": restartCount,
      "lastError": lastError,
      "runtime": "simichat-node-ios-embedded",
      "dependencyMode": "node-mobile-framework",
      "externalProcess": false,
      "appManaged": true,
      "requiresHostNode": false,
      "requiresHostNpx": false,
      "requiresDocker": false,
      "nodeVersion": "18.20.4-mobile",
      "healthUrl": "http://127.0.0.1:\(port)/health",
      "sseUrl": "http://127.0.0.1:\(port)/mcp/sse/simichat-node",
      "scriptPath": scriptPath,
      "stopSupported": false,
    ]
  }

  private func stateName(_ value: Int) -> String {
    switch value {
    case 1: return "starting"
    case 2: return "running"
    case 3: return "crashed"
    default: return "stopped"
    }
  }

  private func applicationSupportDirectory() -> URL {
    FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
      .appendingPathComponent("simichat_bundled_node", isDirectory: true)
  }

  private func prepareRuntimeFiles() throws {
    let runtimeRoot = applicationSupportDirectory()
    try FileManager.default.createDirectory(at: runtimeRoot, withIntermediateDirectories: true)
    let target = runtimeRoot.appendingPathComponent("runtime-server.mjs")
    let candidates = [
      bundle.bundleURL.appendingPathComponent("Frameworks/App.framework/flutter_assets/tools/mcp_runtime/container/runtime-server.mjs"),
      bundle.bundleURL.appendingPathComponent("flutter_assets/tools/mcp_runtime/container/runtime-server.mjs"),
      bundle.url(forResource: "runtime-server", withExtension: "mjs"),
    ].compactMap { $0 }
    guard let source = candidates.first(where: { FileManager.default.fileExists(atPath: $0.path) }) else {
      throw NSError(domain: "SimiChatNodeRuntime", code: 1, userInfo: [NSLocalizedDescriptionKey: "Flutter runtime-server.mjs asset not found"])
    }
    if FileManager.default.fileExists(atPath: target.path) {
      try FileManager.default.removeItem(at: target)
    }
    try FileManager.default.copyItem(at: source, to: target)
    scriptPath = target.path
  }
}

File: android/app/src/main/java/top/simitalk/aichat/SimiChatNodeRuntime.kt
package top.simitalk.aichat

import android.content.Context
import android.content.res.AssetManager
import java.io.File
import java.io.FileOutputStream
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger

/**
 * Owns the Node.js runtime shipped inside the Android APK.
 *
 * This is deliberately a small platform boundary. The MCP server remains a
 * normal Node.js module/script, while libnode.so is started in a dedicated
 * native thread. No host node, npx, Docker, or network download is involved
 * after the APK is installed.
 */
class SimiChatNodeRuntime(private val context: Context) {
    companion object {
        const val CHANNEL = "top.simitalk.aichat/node_runtime"
        const val PORT = 37651
        private const val STATE_STOPPED = 0
        private const val STATE_STARTING = 1
        private const val STATE_RUNNING = 2
        private const val STATE_CRASHED = 3

    }

    private val prepared = AtomicBoolean(false)
    private val restartCount = AtomicInteger(0)
    @Volatile private var nativeLoaded = false
    @Volatile private var nativeLoadError: String? = null
    private val runtimeDirectory: File
        get() = File(context.filesDir, "simichat_node_runtime")
    private val scriptFile: File
        get() = File(runtimeDirectory, "runtime-server.mjs")

    fun start(): Map<String, Any> {
        if (!ensureNativeLoaded()) return info(false)
        prepareAssets()
        if (nativeState() == STATE_CRASHED) {
            restartCount.incrementAndGet()
        }
        val started = nativeStart(
            arrayOf("node", scriptFile.absolutePath),
            runtimeDirectory.absolutePath,
            context.cacheDir.absolutePath,
        )
        // `started` also covers the short STARTING window. Returning it as a
        // positive start acknowledgement lets Dart wait for /health instead
        // of reporting a transient false negative to the MCP layer.
        return info(started || nativeIsRunning())
    }

    fun status(): Map<String, Any> {
        if (!ensureNativeLoaded()) return info(false)
        return info(nativeIsRunning())
    }

    fun stop(): Map<String, Any> {
        // The embedded Node entry point is intentionally process-lifetime
        // scoped. Android destroys the app process and its Node loop together;
        // exposing a fake stop would make the UI report a false state.
        return status() + mapOf("stopSupported" to false)
    }

    private fun info(running: Boolean): Map<String, Any> {
        val loaded = nativeLoaded
        val currentNativeState = if (loaded) nativeState() else STATE_CRASHED
        val exitCode = if (loaded) nativeExitCode() else -1
        return mapOf(
            "running" to running,
            "state" to stateName(currentNativeState),
            "nativeState" to currentNativeState,
            "nativeExitCode" to exitCode,
            "restartCount" to restartCount.get(),
            "lastError" to if (currentNativeState == STATE_CRASHED) {
                nativeLoadError ?: "embedded Node exited with code $exitCode"
            } else {
                ""
            },
            "runtime" to "simichat-node-android-embedded",
            "dependencyMode" to "bundled_nodejs_mobile",
            "externalProcess" to false,
            "appManaged" to true,
            "requiresHostNode" to false,
            "requiresHostNpx" to false,
            "requiresDocker" to false,
            "nodeVersion" to "18.20.4-mobile",
            "abi" to android.os.Build.SUPPORTED_ABIS.firstOrNull().orEmpty(),
            "healthUrl" to "http://127.0.0.1:$PORT/health",
            "sseUrl" to "http://127.0.0.1:$PORT/mcp/sse/simichat-node",
            "scriptPath" to scriptFile.absolutePath,
        )
    }

    private fun ensureNativeLoaded(): Boolean {
        if (nativeLoaded) return true
        synchronized(this) {
            if (nativeLoaded) return true
            return try {
                // Load lazily so an unsupported ABI or a missing native
                // dependency cannot crash the chat UI before the first frame.
                System.loadLibrary("node")
                System.loadLibrary("simichat_node_bridge")
                nativeLoaded = true
                nativeLoadError = null
                true
            } catch (error: Throwable) {
                nativeLoadError = "embedded Node native library unavailable: ${error.message ?: error::class.java.simpleName}"
                false
            }
        }
    }

    private fun stateName(value: Int): String = when (value) {
        STATE_STARTING -> "starting"
        STATE_RUNNING -> "running"
        STATE_CRASHED -> "crashed"
        else -> "stopped"
    }

    private fun prepareAssets() {
        if (prepared.get()) return
        runtimeDirectory.mkdirs()
        copyAssetIfChanged("tools/mcp_runtime/container/runtime-server.mjs", scriptFile)
        prepared.set(true)
    }

    private fun copyAssetIfChanged(assetPath: String, target: File) {
        val marker = File(target.parentFile, ".${target.name}.sha256")
        val bytes = listOf("flutter_assets/$assetPath", assetPath)
            .asSequence()
            .mapNotNull { path ->
                try {
                    context.assets.open(path).use { it.readBytes() }
                } catch (_: java.io.IOException) {
                    null
                }
            }
            .firstOrNull()
            ?: error("Flutter asset not found: $assetPath")
        val digest = java.security.MessageDigest.getInstance("SHA-256")
            .digest(bytes)
            .joinToString("") { "%02x".format(it) }
        if (target.isFile && marker.isFile && marker.readText() == digest) return
        FileOutputStream(target).use { it.write(bytes) }
        marker.writeText(digest)
    }

    private external fun nativeIsRunning(): Boolean

    private external fun nativeState(): Int

    private external fun nativeExitCode(): Int

    private external fun nativeStart(
        arguments: Array<String>,
        workingDirectory: String,
        cacheDirectory: String,
    ): Boolean

}
